### PR TITLE
up_squared: set the flash runner.

### PR DESCRIPTION
**Pull request by chen-png**
set the flash runner of up squared board using misc-flasher according to #20273 and #20659.

Signed-off-by: peng1 chen <peng1.chen@intel.com>

### CODE REVIEW

**Comment by chen-png** [comment]
@nashif Hi, Anas, Could you take a look of this? I add the west/make flash command to up squared board, thus it can run all test cases automatically via sanitycheck.

**Comment by gnuless** [comment]
Nice work.  Just a few thoughts: I wonder if something so specialized belongs in the tree?  On other arches/boards, `make flash` generally can only mean one thing, but here, it's only applicable to one specific setup.  If we do want to include this kind of specialized function, perhaps it could be more fully documented somewhere?

**Review by LeiW000 — APPROVED** [review]

**Comment by chen-png** [comment]
> Nice work. Just a few thoughts: I wonder if something so specialized belongs in the tree? On other arches/boards, `make flash` generally can only mean one thing, but here, it's only applicable to one specific setup. If we do want to include this kind of specialized function, perhaps it could be more fully documented somewhere?

yes, it's a little specified for some boards. Maybe I can add some sections in boards/x86/up_squared/doc/index.rst to introduce this.

**Review by wentongwu — APPROVED** [review]

**Review by maksimmasalski — APPROVED** [review]
LGTM.

File: scripts/west_commands/runners/x86.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Copyright (c) 2017 Linaro Limited.
```
Comment (nashif) [inline scripts/west_commands/runners/x86.py]: There is nothing x86-specific about this method of flashing, please remove the x86 naming and call it tftp or something, you are just copying a file to a directory here.

File: scripts/west_commands/runners/x86_configuration.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+power_off = curl http://admin:1234@192.168.1.20/script?run003=run
+tftp_root_dir = ~/vftp/kernel
```
Comment (nashif) [inline scripts/west_commands/runners/x86_configuration.txt]: Please remove this hardcoding and use command line options.

File: scripts/west_commands/runners/x86_configuration.txt [hunk 3]
```diff
@@ -0,0 +1,3 @@
...
+power_on = curl http://admin:1234@192.168.1.20/script?run001=run
```
Comment (nashif) [inline scripts/west_commands/runners/x86_configuration.txt]: This is very custom and random without any explanation what that is and what lies behind it. Need to think about how to make this generic and remove the need for that, maybe add a command line option that invokes a script that lives out of the tree to reboot the target, without having all of this in the tree.

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/west_commands/runners/x86_configuration.txt (hunk 3 above)
Comment (chen-png) [reply scripts/west_commands/runners/x86_configuration.txt]: @nashif  yes, I can add an argument behind west flash command using this runner's do_add_parser function, whick will set the script location, in this way,  the west flash --option works well.
So if I want to run sanitycheck, I also have to add this option to sanitycheck via parser.add_argument in sanitycheck script, and pass the added option to the final west flash command in devicehandler function. I have to modify two place in the sanitycheck script, do you think this is an reasonable way?

**Comment by chen-png** [comment]
@nashif Hi, Anas, I modified it by adding a commond option in sanitycheck script, Could you take a look of it?

File: scripts/sanitycheck [hunk 4]
```diff
@@ -689,6 +689,11 @@ class DeviceHandler(Handler):
             if options.west_flash != []:
                 command.append('--')
                 command.extend(options.west_flash.split(','))
+            # This argument is used for boards like up_squared
+            # to run sanitycheck script.
+            if options.tftp_config is not None:
```
Comment (nashif) [inline scripts/sanitycheck]: This does not belong into sanitycheck, all of this needs to be supported in west and the flasher and provided to sanitycheck as an option via --west-flash

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/sanitycheck (hunk 4 above)
Comment (chen-png) [reply scripts/sanitycheck]: Do you mean that I don't need to add another option, I should just use the existing --west-flash option?
for now, I can use west flash --option well by adding support to tftp runner. yes, I can just add the option behind --west-flash, don't need to modify sanitycheck for other options.

File: scripts/west_commands/runners/tftp.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    @classmethod
+    def create(cls, cfg, args):
+        return TftpBinaryRunner(cfg, args.tftp_config)
+
+    def do_run(self, command, **kwargs):
+        # test case image
+        strip_file = self.cfg.elf_file[:-4] + ".strip"
+        # invoke the external script set by tftp_config
+        # which will copy the iamge and complete the pxe boot
+        os.system("%s %s" %(self.tftp_config, strip_file))
```
Comment (mbolivar) [inline scripts/west_commands/runners/tftp.py]: I don't understand what's going on here, sorry. Nothing in this file runs tftp. What's going on?

The comment above says this script "use the network switch to control the power of board, and the board can download the necessary image from the tftp root dir via pxe boot".

But this is running an arbitrary executable set up via the runner arguments in the board file, so there's nothing tftp specific here. And the runner arguments in the up_squared board file in this commit don't specify any command to run, so I don't see how this can possibly work. What am I missing?

Comment (chen-png) [reply scripts/west_commands/runners/tftp.py]: @mbolivar the executable set up file will be invoked via the --west-flash option in sanitycheck script, not set by up squared board file. We can use the sanitycheck --west-falsh="--tftp_config=filepath" to set the executable file path which will run the power off, copy image and power on command in it.
in this way, the tftp config file is out of the zephyr source tree, and you can realize your own pxe boot way for up board in the tftp config file, you just have to pass the file path to sanitycheck script.

Comment (mbolivar) [reply scripts/west_commands/runners/tftp.py]: That doesn't address the concern that "nothing in this file runs tftp"

Comment (chen-png) [reply scripts/west_commands/runners/tftp.py]: @mbolivar In order to run all test cases automatically, we set up the up board boot by network, and set up a pxe server, if the board power on, it will download the grub boot loader image and test case image from the tftp root dir in pxe server. So the tftp there means a directory contains the test case image build in pxe server.

The argument tftp_config actually is a shell script which can control the pxe boot of up squared board remotely. In this script, there are three command,
1. power on: curl http://admin:1234@192.168.1.20/script?run003=run, we use a network switch to control the power of up board.
2. copy image: cp $0 ~/vftp/kernel, the $0 is the argument strip_file the test case image named zephyr.strip build in  pxe server, when the board power on, it will download those images automatically.
3. power on: curl http://admin:1234@192.168.1.20/script?run001=run

but maybe you can realize your own way to control the power of up board, so I just pass the file path to sanitycheck.

Comment (mbolivar) [reply scripts/west_commands/runners/tftp.py]: I understand that you're passing the script an argument that does run tftp. However, this runner is named `tftp`, but nothing at all in the file is using tftp. It just takes an arbitrary shell command and passes it the elf file.

It's not appropriate to name this runner tftp, even if that is how you happen to be using it downstream when you run sanitycheck.

If you want to name a runner tftp and get it merged upstream, I would expect it to start up a tftp server at port 69 and make a file available to at least one client, along with some options it would take from the board file for how to reset the client and start it downloading, decide when the client is done, etc.

**Review by mbolivar — CHANGES_REQUESTED** [review]
adding -1 since this has other approvals

**Comment by chen-png** [comment]
@mbolivar Could you take a review of this PR? i named the runner as "up squared".

**Comment by chen-png** [comment]
@nashif  Hi anas, Could you take a review of this?

File: boards/x86/up_squared/board.cmake [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (nashif) [inline boards/x86/up_squared/board.cmake]: this needs to be in a separate commit.

File: scripts/west_commands/runners/up_squared.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    @classmethod
+    def capabilities(cls):
+        # only have to support the flash command.
+        return RunnerCaps(commands={'flash'})
+
+    @classmethod
+    def do_add_parser(cls, parser):
+        # using the pxe-boot.sh script to achieve the goal copying
+        # test case iamge and grub image to board
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: image

File: scripts/west_commands/runners/up_squared.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+'''Runner for flashing on up squared board via pxe boot.'''
+
+from runners.core import ZephyrBinaryRunner, RunnerCaps
+
+
+class UpSquaredBinaryRunner(ZephyrBinaryRunner):
+    '''Runner front-end for Up Squared.'''
+    # This is an special flash command for up squared board.
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: This is a special

File: scripts/west_commands/runners/up_squared.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    # It will invoke the pxe_boot_sh script set by '--pxe-boot'
+    # option to complete the test case image downloading for
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: what test case? this should work for everything, not only testcases...

File: scripts/west_commands/runners/up_squared.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        # test case iamge and grub image to board
+        parser.add_argument('--pxe-boot', required=True,
+                            help='pxe boot control command')
+
+    @classmethod
+    def create(cls, cfg, args):
+        return UpSquaredBinaryRunner(cfg, args.pxe_boot)
+
+    def do_run(self, command, **kwargs):
+        if self.pxe_boot_sh is None:
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: if not self.pxe_boot_sh:

File: scripts/west_commands/runners/up_squared.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            raise ValueError('Cannot flash; --pxe-boot not given.')
+        # test case image "zephyr.strip"
+        strip_file = self.cfg.elf_file[:-4] + ".strip"
+        # run pxe_boot_sh script
+        cmd = [self.pxe_boot_sh, strip_file]
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: You need to define the interface of that script somewhere, so whoever is going to implement it knows what to expect from west and how to write this script.
I also do not see the script, an example or a template for the script that could work with this runner would be very useful.

File: scripts/west_commands/runners/up_squared.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
+# Copyright (c) 2017 Linaro Limited.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''Runner for flashing on up squared board via pxe boot.'''
```
Comment (nashif) [inline scripts/west_commands/runners/up_squared.py]: This really needs to be abstracted and made generic, there is nothing related to the up_squared board here and there is also nothing related to PXE. You are basically just calling a script that can do lots of things, so this runner need to be made more generic. Do not make it appear as it only needed for this specific board or PXE. Basically It can be called ScriptedRunner with an option pointing to a script without having to mention pxe or the up_squared board.

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/west_commands/runners/up_squared.py (hunk 12 above)
Comment (mbolivar) [reply scripts/west_commands/runners/up_squared.py]: I agree that this trend of adding runners which only work for one board (like intel_s1000) should not continue.

If this can't be turned into a generic runner that makes using tftp easier (which would be my preference), I agree that having a generic runner -- name something like CommandRunner because you might not be calling a 'script' -- which requires a binary and some arguments to run is a good way forward.

This would make it easier for special-purpose out of tree boards in addition to up_squared.

I'm thinking something like this for do_add_parser (untested):

```
parser.add_argument('command', help='command to run')
parser.add_argument('extra_args', nargs='*', help='additional command line arguments to pass to command')
```

Then just do `subprocess.check_call(list(args.command) + extra_args)` from do_run or so.

File: scripts/west_commands/runners/up_squared.py (hunk 11 above)
Comment (chen-png) [reply scripts/west_commands/runners/up_squared.py]: I will add a section in the boards/x86/up_squared/doc/index.rst to introduce this, there will be a script template.

File: boards/x86/up_squared/board.cmake (hunk 6 above)
Comment (chen-png) [reply boards/x86/up_squared/board.cmake]: fixed it.

File: scripts/west_commands/runners/up_squared.py (hunk 12 above)
Comment (chen-png) [reply scripts/west_commands/runners/up_squared.py]: I named it as GenericCommandRunner, and it will invoke the command set by the 'command' option.
the needed 'zephyr.strip' is just in the build dir after the west build operation, it is aleready in the parsed cfg argument,  there is no need for another arguments.

File: scripts/west_commands/runners/generic_command.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+'''Runner for flashing with external command option.'''
+
+from runners.core import ZephyrBinaryRunner, RunnerCaps
+
+
+class GenericCommandRunner(ZephyrBinaryRunner):
+    # This is a special flash command which uses the given
+    # 'command' option to achieve the goal copying test
+    # case image and grub boot image to board
```
Comment (mbolivar) [inline scripts/west_commands/runners/generic_command.py]: This is still mixing up things that are specific to your board into a runner that has nothing to do with it:

- this comment which mentions grub
- the git shortlog which says the area is "arch/x86" despite only editing files in "scripts/west_commands"
- assuming that something has created a `zephyr.strip` file which must be passed to the command

It also doesn't include the suggestion that the user should be able to pass arbitrary commands at the command line.

I think the best way to get this PR merged is if I provide you with some infrastructure which you can use for your board. Let's start here if we can: https://github.com/zephyrproject-rtos/zephyr/pull/20273

If we can agree on the infrastructure you need for up_squared and merge that, then we can come back to this PR with a commit that just touches the up_squared board.cmake.

**Review by mbolivar — CHANGES_REQUESTED** [review]

File: scripts/west_commands/runners/generic_command.py (hunk 13 above)
Comment (chen-png) [reply scripts/west_commands/runners/generic_command.py]: it's great.

Comment (chen-png) [reply scripts/west_commands/runners/generic_command.py]: touched a board.cmake file to set flash runner for up squared according to #20273, is it ok?

**Review by mbolivar — APPROVED** [review]

**Comment by mbolivar** [comment]
@nashif up to you what you want to do in mainline. I'm fine with this even though I don't get it

**Comment by chen-png** [comment]
> So ... I admit I don't get why you want this upstream.

because I want to run all test cases on up squared board automatically using sanitycheck script.

up squared board don't support west flash command to run test cases before, so i want to add a runner for it using nerwork and pxe to download image and run test case, then i can just use the sanitycheck to run all test cases on it.

File: boards/x86/up_squared/board.cmake [hunk 14]
```diff
@@ -0,0 +1,4 @@
...
+
+board_set_flasher_ifnset(misc-flasher)
+board_finalize_runner_args(misc-flasher)
```
Comment (mbolivar) [inline boards/x86/up_squared/board.cmake]: This is very interesting... I guess you're calling sanitycheck with the complete path to your script and all its arguments?

I'm not sure why we need to involve west in this, perhaps sanitycheck can just take an option for a custom flash command.

But to keep west involved, I suggested in another PR that we could have misc-flasher set some environment variables to contain at least the zephyr build directory, if not also the bin / hex / elf files inside it. Will that help you?

Comment (chen-png) [reply boards/x86/up_squared/board.cmake]: yes, i call the sanitycheck with my scripts' path and needed arguments. 
becasue sanitycheck has a --west-flash option which can be used to pass my scripts and arguments.

'custom flash'? do you mean the 'make flash'? if i use the make flash, i can't pass my arguments, because the '--west-flash' option is only for the west flash command.

great. if  i know the path of zephyr build directory, i can use my script to search for the test image path in it.
for now, the problem is that I can't determine the test image path when i run sanitycheck, so if i know the build directory, it will be fine.

**Comment by chen-png** [comment]
@nashif with the PR #20273 and #20659 merged, I think this is ok. Can this be merged?

**Review by nashif — APPROVED** [review]

**Comment by chen-png** [comment]
I think this is ok to be merged now. Does anyone help to merge it?
